components/brave_wallet_ui/common/hardware/ledgerjs/sol_ledger_bridge_keyring.ts

```diff
@@ -0,0 +1,116 @@
+import TransportWebHID from '@ledgerhq/hw-transport-webhid'
+import Transport from '@ledgerhq/hw-transport'
+import Sol from '@ledgerhq/hw-app-solana'
+import { BraveWallet } from '../../../constants/types'
+import { LedgerSolanaKeyring } from '../interfaces'
+import { HardwareVendor } from '../../api/hardware_keyrings'
+import {
+  GetAccountsHardwareOperationResult,
+  HardwareOperationResult,
+  SignHardwareMessageOperationResult,
+  SignHardwareTransactionOperationResult
+} from '../types'
+
+import { hardwareDeviceIdFromAddress } from '../hardwareDeviceIdFromAddress'
+
+export default class SolanaLedgerKeyring implements LedgerSolanaKeyring {
+  private app?: Sol
+  private transport?: Transport
+  private deviceId: string
+
+  coin = (): BraveWallet.CoinType => {
+    return BraveWallet.CoinType.SOL
+  }
+
+  type = (): HardwareVendor => {
+    return LEDGER_HARDWARE_VENDOR
+  }
+
+  getAccounts = async (from: number, to: number): Promise<GetAccountsHardwareOperationResult> => {
+    const unlocked = await this.unlock()
+    if (!unlocked.success || !this.app) {
+      return unlocked
+    }
+    from = (from < 0) ? 0 : from
+    const sol: Sol = this.app
+    const accounts = []
+    for (let i = from; i <= to; i++) {
+      const path = this.getPathForIndex(i)
+      const address = await sol.getAddress(path)
+      accounts.push({
+        address: '',
+        addressBytes: address.address,
+        derivationPath: path,
+        name: this.type(),
+        hardwareVendor: this.type(),
+        deviceId: this.deviceId,
+        coin: this.coin()
+      })
+    }
+    return { success: true, payload: [...accounts] }
+  }
+
+  isUnlocked = (): boolean => {
+    return this.app !== undefined && this.deviceId !== undefined
+  }
+
+  makeApp = async () => {
+    this.transport = await TransportWebHID.create()
+    this.app = new Sol(this.transport)
+  }
+
+  unlock = async (): Promise<HardwareOperationResult> => {
+    if (this.isUnlocked()) {
+      return { success: true }
+    }
+
+    if (!this.app) {
+      await this.makeApp()
+    }
+
+    if (this.app && !this.deviceId) {
+      const sol: Sol = this.app
+      this.transport?.on('disconnect', this.onDisconnected)
+      const zeroPath = this.getPathForIndex(0)
+      const address = (await sol.getAddress(zeroPath)).address
+      this.deviceId = await hardwareDeviceIdFromAddress(address)
+    }
+    return { success: this.isUnlocked() }
+  }
+
+  signPersonalMessage (path: string, address: string, message: string): Promise<SignHardwareMessageOperationResult> {
+    throw new Error('Method not implemented.')
+  }
+
+  signTransaction = async (path: string, rawTxBytes: Buffer): Promise<SignHardwareTransactionOperationResult> => {
+    try {
+      const unlocked = await this.unlock()
+      if (!unlocked.success || !this.app) {
+        return unlocked
+      }
+      const sol: Sol = this.app
+      const signed = await sol.signTransaction(path, rawTxBytes)
```

Comment:
can we add a check to make sure that the signed bytes received from the HW wallet match the bytes we sent to it?

Reply:
I believe `signed` in this case is only the signature, and doesn't include the transaction bytes. The transaction bytes are kept in wallet state and then recombined with the signature before submitting to the network here https://github.com/brave/brave-core/pull/13734/files/19dbc9929a5f909ffdc52be0344bf740283ebc8a#diff-2042ad5dd3965e1a3c1a30ffe42f6610b0975da8a8a89f9dca00fd92549901f4R511-R512.

So I believe in this case if the hardware wallet returned a signature for a different transaction it would fail validation when when submitted to the network

Reply:
This is the one I meant to comment on. Thanks for explaining this. The design should avoid this concern